Runtime intrinsics: fix fpext and fptrunc behaviour on Float16/BFloat16

xal-0 opened the pull request:
This makes two changes `fpext` and `fptrunc` to match the behaviour specified in their error strings:
-`fpext` works when converting from Float16 => Float16,
-`fptrunc` is prevented from truncating Float16 => Float16

Both are re-written to make it explicit what conversions are possible, and how they are done.

Closes #57130.

xal-0 committed:
Add additional fpext/fptrunc tests for Float16/BFloat16

xal-0 committed:
Enforce bitsize restrictions on fptrunc/fpext (fixes #57130)

xal-0 commented:
I'm still not sure what the right thing to do is when doing `fpext`/`fptrunc` to/from BFloat16. Currently, it is allowed when "output bits >= input bits" holds for `fpext`, and when "input bits < output bits" for `fptrunc`.  This has the weird effect of losing information on `fpext`:

```
julia> using .Core.Intrinsics
julia> using .Core: BFloat16
julia> fpext(Float16, fpext(BFloat16, Float16(12.3)))
Float16(12.31)
```

oscardssmith commented:
I believe the correct behavior would be for fpext to allow going from Float16 or Bfloat16 to Float32/Float64 but not from Float16 to BFloat16 (or from Bfloat16 to Float16). I think that fpext is supposed to only allow transitions from  src to dest if all possible src values are losslessly encodable in dest.

xal-0 commented:
LLVM seems to prohibit `fpext` and `fptrunc` with input size = output size:
```
llc: error: llc: fpstuff.ll:2:14: error: invalid cast opcode for cast from 'float' to 'float'
  %y = fpext float %x to float
             ^

llc: error: llc: fpstuff.ll:70:16: error: invalid cast opcode for cast from 'half' to 'bfloat'
  %y = fptrunc half %x to bfloat
               ^

llc: error: llc: fpstuff.ll:75:14: error: invalid cast opcode for cast from 'half' to 'bfloat'
  %y = fpext half %x to bfloat
             ^
```

maleadt reviewed (approved):
LGTM (superficially at least, it's been a while since I looked at this code).

For BFloat16s.jl it'd be fine to reject fpext/fptrunc between f16 and bf16, it currently should go through f32 anyway.

vtjnash commented on src/runtime_intrinsics.c:
```diff
@@ -1629,32 +1621,85 @@ cvt_iintrinsic(LLVMUItoFP, uitofp)
+    fptrunc_convert(float32, bfloat16);
+    fptrunc_convert(float64, bfloat16);
+    fptrunc_convert(float64, float32);
+    else
+        jl_error("fptrunc: runtime floating point intrinsics are not implemented for bit sizes other than 16, 32 and 64");
```
Since you are changing the permitted combination of behaviors, you will also need to update Compiler/src/tfuncs.jl and src/intrinsics.cpp to match these new rules, so that they all agree on exactly which errors are thrown and for what cases they can occur

xal-0 committed:
Prohibit fpext with input = output bitsize to match LLVM

We also throw an exception when using fpext/fptrunc on integer types, because
with the addition of BFloat16, there is no unambiguous floating point format
with which to interpret the input when it is 16 bits wide.

xal-0 committed:
Update effect inference, codegen for new fpext/fptrunc behaviour

BFloat16 tests that trigger codegen are disabled for now, pending an LLVM fix.

xal-0 commented:
I decided to leave the incorrect handling of NaNs for later, since what to do is not entirely clear (#49353).

xal-0 committed:
Add tests for fptrunc on NaNs, fix float_to_half NaN handling

double_to_half, float/double_to_bfloat TODO

vtjnash commented on Compiler/src/tfuncs.jl:
```diff
@@ -2454,6 +2454,9 @@ const _SPECIAL_BUILTINS = Any[
 ]
 
+# Types compatible with fpext/fptrunc
+const _FLOAT_TYPES = Any[Core.BFloat16, Float16, Float32, Float64]
```
```suggestion
const CORE_FLOAT_TYPES = Union{Core.BFloat16, Float16, Float32, Float64}
```

vtjnash commented on Compiler/src/tfuncs.jl:
```diff
@@ -2867,6 +2870,17 @@ function intrinsic_exct(𝕃::AbstractLattice, f::IntrinsicFunction, argtypes::V
+
+        # fpext and fptrunc have further restrictions on the allowed types.
+        if f === Intrinsics.fpext &&
+            !(ty in _FLOAT_TYPES && xty in _FLOAT_TYPES && Core.sizeof(ty) > Core.sizeof(xty))
```
```suggestion
            !(ty <: CORE_FLOAT_TYPES && xty <: CORE_FLOAT_TYPES && Core.sizeof(ty) > Core.sizeof(xty))
```

xal-0 committed:
Rename _FLOAT_TYPES to CORE_FLOAT_TYPES, make it a Union{...}

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

oscardssmith commented on test/intrinsics.jl:
```diff
@@ -215,18 +281,60 @@ end
+    x = (sign&1)<<31 | (exp&((1<<8)-1))<<23 | sig&((1<<23)-1)
+    return reinterpret(Float32, UInt32(x))
+end
+function f64(sign, exp, sig)
+    x = (sign&1)<<31 | (exp&((1<<11)-1))<<52 | sig&((1<<52)-1)
+    return reinterpret(Float64, UInt64(x))
+end
```
This is wrong on 32 bit systems (you want `x = Int64(sign&1)<<31 | Int64(exp&((1<<11)-1))<<52 | sig&((Int64(1)<<52)-1)`

xal-0 committed:
Fix test/intrinsics.jl when Int = Int32
